### ISSUE TITLE
h264_mediacodec decoding far too slow, slower than software dekoding

### ISSUE THREAD

**Issue by geminixdev** [valbok/QtAVPlayer#303]
Meanwhile it's clear that the issues I had (https://github.com/valbok/QtAVPlayer/issues/299 ), and which seemed to have been solved by using a different set of ffmpeg libraries, where in reality issues with the decoding performance of h264_mediacodec. 

The difference in the ffmpeg libraries I used where if they support mediacodec or not.

**With mediacodec decoding** is much too slow, far away from 25 fps, less than half of that, while decoding in bursts, some frames are coming quickly and other frames very very late, 100 - 400 ms. (As described in #299 ).

**With software decoding** the performance on arm64-v8a devices is good enough for HD (720p) in 25 fps. However on many armeabi-v7a devices the software decoding performance is not good enough for HD. 

_So the question is, how to get h264_mediacodec to perform as expected, and faster than software decoding?_

Ffmpeg libraries tested where Javernaut built ffmpeg 5.1.3 with and without mediacodec (https://github.com/Javernaut/ffmpeg-android-maker/issues/76 ), and the prebuilt ones from ffmpegKit (which include mediacodec).

Searching the internet for days I found reports of others having the same problem, that software decoding is faster than h264_mediacodec, but no solution, only problem reports.

The problem is visible and easy to reproduce and test using the **qml_video example**. Did anyone run that code with h264_mediacodec and good fps performance?

**Comment by geminixdev** [comment]
while examining the log output of decoding with h264_mediacodec, I realized this:

The time between 2 QAVVideoFrames, as delivered by QAVPlayer::videoFrame, measured with QElapsedTimer, varies dramatically, between 1 ms and about 1100 ms.
When the elapsed time is very high, then the PTS increase is not the expected 3600 (25 fps), but 10800, 14400, 18000. Always a multiple of 3600.

The conclusion is that there are frames lost, 2 or 3 or even 4 frames. The video quality, when playing, looks also exactly like that, (reference) frames missing.

See this log extract, showing elapsed tme and PTS increase:

```
I Player  : [20 2:37:48.323 +07 I]  Video  19 ms   frame()->pts 6479708227 pkt_pts incr: 3600
I Player  : [20 2:37:48.323 +07 I]  Video   0 ms   frame()->pts 6479711827 pkt_pts incr: 3600
I Player  : [20 2:37:48.349 +07 I]  Video  26 ms   frame()->pts 6479715427 pkt_pts incr: 3600
I Player  : [20 2:37:48.385 +07 I]  Video  36 ms   frame()->pts 6479719027 pkt_pts incr: 3600
I Player  : [20 2:37:48.703 +07 I]  Video 318 ms   frame()->pts 6479737027 pkt_pts incr: 18000
I Player  : [20 2:37:48.716 +07 I]  Video  13 ms   frame()->pts 6479740627 pkt_pts incr: 3600
I Player  : [20 2:37:48.717 +07 I]  Video   1 ms   frame()->pts 6479744227 pkt_pts incr: 3600
I Player  : [20 2:37:48.732 +07 I]  Video  15 ms   frame()->pts 6479747827 pkt_pts incr: 3600
I Player  : [20 2:37:48.867 +07 I]  Video 135 ms   frame()->pts 6479762227 pkt_pts incr: 14400
I Player  : [20 2:37:48.903 +07 I]  Video  36 ms   frame()->pts 6479765827 pkt_pts incr: 3600
I Player  : [20 2:37:48.949 +07 I]  Video  45 ms   frame()->pts 6479769427 pkt_pts incr: 3600
I Player  : [20 2:37:48.983 +07 I]  Video  35 ms   frame()->pts 6479773027 pkt_pts incr: 3600
I Player  : [20 2:37:49.150 +07 I]  Video 167 ms   frame()->pts 6479787427 pkt_pts incr: 14400
I Player  : [20 2:37:49.183 +07 I]  Video  33 ms   frame()->pts 6479791027 pkt_pts incr: 3600
I Player  : [20 2:37:49.216 +07 I]  Video  32 ms   frame()->pts 6479794627 pkt_pts incr: 3600
I Player  : [20 2:37:49.303 +07 I]  Video  88 ms   frame()->pts 6479805427 pkt_pts incr: 10800
I Player  : [20 2:37:49.350 +07 I]  Video  47 ms   frame()->pts 6479809027 pkt_pts incr: 3600
I Player  : [20 2:37:49.383 +07 I]  Video  33 ms   frame()->pts 6479812627 pkt_pts incr: 3600
I Player  : [20 2:37:49.431 +07 I]  Video  48 ms   frame()->pts 6479816227 pkt_pts incr: 3600
I Player  : [20 2:37:49.466 +07 I]  Video  34 ms   frame()->pts 6479819827 pkt_pts incr: 3600
I Player  : [20 2:37:49.515 +07 I]  Video  49 ms   frame()->pts 6479823427 pkt_pts incr: 3600
I Player  : [20 2:37:49.969 +07 I]  Video 454 ms   frame()->pts 6479845027 pkt_pts incr: 21600
I Player  : [20 2:37:49.969 +07 I]  Video   1 ms   frame()->pts 6479848627 pkt_pts incr: 3600
I Player  : [20 2:37:49.985 +07 I]  Video  16 ms   frame()->pts 6479852227 pkt_pts incr: 3600
I Player  : [20 2:37:50.081 +07 I]  Video  96 ms   frame()->pts 6479863027 pkt_pts incr: 10800
I Player  : [20 2:37:50.082 +07 I]  Video   1 ms   frame()->pts 6479866627 pkt_pts incr: 3600
I Player  : [20 2:37:50.082 +07 I]  Video   0 ms   frame()->pts 6479870227 pkt_pts incr: 3600
I Player  : [20 2:37:50.083 +07 I]  Video   1 ms   frame()->pts 6479873827 pkt_pts incr: 3600
I Player  : [20 2:37:50.115 +07 I]  Video  32 ms   frame()->pts 6479877427 pkt_pts incr: 3600
I Player  : [20 2:37:50.163 +07 I]  Video  48 ms   frame()->pts 6479881027 pkt_pts incr: 3600
I Player  : [20 2:37:50.501 +07 I]  Video 338 ms   frame()->pts 6479899027 pkt_pts incr: 18000
I Player  : [20 2:37:50.515 +07 I]  Video  14 ms   frame()->pts 6479902627 pkt_pts incr: 3600
I Player  : [20 2:37:50.515 +07 I]  Video   0 ms   frame()->pts 6479906227 pkt_pts incr: 3600
I Player  : [20 2:37:50.530 +07 I]  Video  15 ms   frame()->pts 6479909827 pkt_pts incr: 3600
I Player  : [20 2:37:50.813 +07 I]  Video 283 ms   frame()->pts 6479927827 pkt_pts incr: 18000
I Player  : [20 2:37:50.814 +07 I]  Video   1 ms   frame()->pts 6479931427 pkt_pts incr: 3600
I Player  : [20 2:37:50.829 +07 I]  Video  15 ms   frame()->pts 6479935027 pkt_pts incr: 3600
I Player  : [20 2:37:50.931 +07 I]  Video 102 ms   frame()->pts 6479945827 pkt_pts incr: 10800
I Player  : [20 2:37:50.932 +07 I]  Video   1 ms   frame()->pts 6479949427 pkt_pts incr: 3600
```

This is happening only when using h264_mediacodec, not when using software dekoding:

```
D Player  : [20 2:37:45.859 +07 D]  Available pixel formats:
D Player  : [20 2:37:45.859 +07 D]     mediacodec : AVPixelFormat( 165 )
I OMXClient: IOmx service obtained
D MediaCodec: flushMediametrics
W ACodec  : [OMX.Exynos.avc.dec] setting HDRStaticInfo failed even though codec advertises support
W ACodec  : [OMX.Exynos.avc.dec] getting HDRStaticInfo failed even though codec advertises support
D Player  : [20 2:37:45.941 +07 D]  FFmpeg: [h264_mediacodec @ 0x7b8d6350b0] Output crop parameters top=0 bottom=575 left=0 right=1023, resulting dimensions width=1024 height=576
04-20 02:37:45.941 20362 16836 D Player  :
D Player  : [20 2:37:45.942 +07 D]  FFmpeg: [h264_mediacodec @ 0x7b8d6350b0] MediaCodec started successfully: codec = OMX.Exynos.avc.dec, ret = 0
04-20 02:37:45.942 20362 16836 D Player  :
I Player  : [20 2:37:45.948 +07 I]  mediaStatusChanged LoadedMedia , State: PlayingState
I Player  : [20 2:37:45.949 +07 I]  Video streams: 1
I Player  : [20 2:37:45.949 +07 I]  [ 0 ] QMap() ---current
I Player  : [20 2:37:45.950 +07 I]  Audio streams: 1
I Player  : [20 2:37:45.950 +07 I]  [ 1 ] QMap() ---current
W ACodec  : [OMX.Exynos.avc.dec] getting HDRStaticInfo failed even though codec advertises support
I libOpenSLES: Emulating old channel mask behavior (ignoring positional mask 0x3, using default mask 0x3 based on channel count of 2)
D Player  : [20 2:37:45.993 +07 D]  FFmpeg: [h264_mediacodec @ 0x7b8d6350b0] Output MediaFormat changed to {crop-right=1023, vendor.rtc-ext-dec-caps-vt-driver-version.number=180315, vendor.rtc-ext-dec-low-latency.enable=0, color-format=21, slice-height=576, image-data=java.nio.HeapByteBuffer[pos=0 lim=104 cap=104], mime=video/raw, vendor.sec-ext-dec-compressed-color-format=0, stride=1024, color-standard=1, color-transfer=3, crop-bottom=575, crop-left=0, width=1024, color-range=2, crop-top=0, height=576}
04-20 02:37:45.993 20362 16847 D Player  :
D Player  : [20 2:37:45.993 +07 D]  FFmpeg: [h264_mediacodec @ 0x7b8d6350b0] Output crop parameters top=0 bottom=575 left=0 right=1023, resulting dimensions width=1024 height=576
```

Where are these missing frames getting lost, and how to avoid?

**Comment by geminixdev** [comment]
I found a hint where mediacodec behaves weird and might need a special treatment, otherwise frames get lost:

https://programmer.ink/think/61d3158205699.html

**Comment by valbok** [comment]
> The conclusion is that there are frames lost, 2 or 3 or even 4 frames. The video quality, when playing, looks also exactly like that, (reference) frames missing.

This can be also when pts for video and audio is not synced. To confirm, need to make sure that pts even exists, might be bug inside QtAVPlayer. 

Frames are skipped if audio decoder does not provide correct pts.

**Comment by valbok** [comment]
Interesting that tried on my old env with old ffmpeg

**Comment by valbok** [comment]
It looks like a bug since the decoder might return several frames from one packet, checking this.

**Comment by geminixdev** [comment]
> It looks like a bug since the decoder might return several frames from one packet, checking this.

I have it kind of working now, no more frames lost, picture is clear and without errors now. Nevertheless it is still overall too slow, on most devices not better than software decoding.

In order to make it work, I had to make quite a few dirty changes in the elegant setup of QAVFrameCodec, QAVDemuxer and QAVPlayer. The flow required by MediaCodec does not fit into the usual decoding workflow of ffmpeg. The blogpost I mentioned (https://programmer.ink/think/61d3158205699.html) describes the solution correctly, but the ffmpeg oriented way all is constructed in QtAVPlayer does not make that "ffmpeg workflow violation", as described there, easy to implement. But it works. 

- On one Android box, which cannot play Full HD with software encoding, now with MediaCodec it works. Not perfect, it seems to be at the limit, but it is watchable. 
- On other devices it is about the same as software decoding.

Overall performance is still not good enough, there seems to be somewhere a bottleneck. I'm still analyzing and searching for it. Work in progress, will update more when I see more.

**Comment by geminixdev** [comment]
> > The conclusion is that there are frames lost, 2 or 3 or even 4 frames. The video quality, when playing, looks also exactly like that, (reference) frames missing.
> 
> This can be also when pts for video and audio is not synced. To confirm, need to make sure that pts even exists, might be bug inside QtAVPlayer.
> 
> Frames are skipped if audio decoder does not provide correct pts.

I had logged the PTS from Audio and Video, as it gets delivered, and then also at it gets submitted to mediacodec and as it returns. The audio PTS was always as expected, but inside mediacodec there will be frames lost, if it is not done in the order the blogpost explains.

Simplified:
- you must **always** try to receive a frame before you can send a packet to mediacodec. 
- And even then you cannot always send successfully, so you must **keep that packet and resend it** next time. 

In `bool QAVFrameCodec::decode(const AVPacket *pkt, AVFrame *frame) const`, this cannot get implemented in a function returning bool, because there are 4-5 different results possible, which need to get considered:

- a frame was received, process it. Packet was NOT successfully sent though.
- a frame was received, process it. Packet was sent too.
- no frame received, instead AVERROR(EAGAIN). Now the packet was sent.
- no frame received, instead AVERROR(EAGAIN). Packet was NOT successfully sent though
- no frame received, no AVERROR(EAGAIN). Packet was NOT successfully sent though.

Once it is ensured that _not sent packets_ don't get lost, but reused next time, then the decoding works without any lost frames, PTS values are correct and as expected, no jumps or drops, and can get used to sync audio and video, and all resulting decoded images are clean.

**Comment by geminixdev** [comment]
> Overall performance is still not good enough, there seems to be somewhere a bottleneck. I'm still analyzing and searching for it. Work in progress, will update more when I see more.

Found it. What is slowing it down is the syncing to the fps rate in `QAVPlayerPrivate::doPlayStep`. 

After taking off that brake it speeds through with around 1 ms decoding time per frame. I moved the video frame syncing after the decoding, and all good now, I'm happy.

**Comment by valbok** [comment]
> Once it is ensured that not sent packets don't get lost, but reused next time, then the decoding works without any lost frames, PTS values are correct and as expected,

**Comment by geminixdev** [comment]
> Does it mean from your findings, one packet produces exactly one frame? It was just rejected to receive?

Yes. But there is a delay. The first frame out comes after pushing 4 or 5 packets in.

`avcodec_receive_frame` tells you by returning AVERROR(EAGAIN) that it needs you to send more packets. Nevertheless I noticed that you can send another packet also when a frame was successfully received. 

And `avcodec_send_packet` returns 0 when sent successfully. If not returning 0 you need to resend the packet. but always run a `avcodec_receive_frame` before sending another packet. This is key for it to work.

From my checks of PTS into `avcodec_send_packet` and PTS out of `avcodec_receive_frame`, there was for every PTS in a PTS out, and no additional PTS values out. There was just this several frames delay between in and out.

Here is the working decode function:

```
// result 1: a frame was received, process it. Packet was NOT sent though. return ? => Frame OK, prepend packet => return 1
// result 1a: a frame was received, process it. Packet was sent too. return ? => Frame OK, packet processed => return 2
// result 2: no frame received, instead AVERROR(EAGAIN). Now the packet was sent. return ? => NO Frame, packet processed.  => return 0
// result 3: no frame received, instead AVERROR(EAGAIN). Packet was NOT sent though, because of an error. return ? =>  NO Frame, prepend packet  => return -1
// result 4: no frame received, no AVERROR(EAGAIN). Packet was NOT sent though. return ? => NO Frame, prepend packet  => return -1
int QAVFrameCodec::decode_mediacodec(const AVPacket *pkt, AVFrame *frame) const
{
    Q_D(const QAVCodec);
    if (!d->avctx)
        return -1;

    int result = -1;
    bool frame_received = false;
    bool packet_sent = false;
    int ret2 = 999;

    int ret = avcodec_receive_frame(d->avctx, frame);

    if (ret == 0) {
        frame_received = true;
    }

    if (frame_received || ret == AVERROR(EAGAIN)) {
        ret2 = avcodec_send_packet(d->avctx, pkt);

        if (ret2 == 0) {
            packet_sent = true;
        }
    }

    if (frame_received && packet_sent)
        result = 2;
    else if (frame_received)
        result = 1;
    else if (packet_sent)
        result = 0;

    return result;
}
```

**Comment by geminixdev** [comment]
Issues I still have, on which I'm still working. May be you have an idea how to approach them:

- I'm testing on several devices. Although mediacodec works on all of them, there are some armeabi-v7a devices where the mediacodec decoding is not faster, possibly even slower than software decoding. I don't see an error or hint for a reason in logcat.
- On another armeabi-v7a device the GUI thread seems to get overloaded when playing Full HD, which is visible in the video not playing smoothly, despite decoding being fast. HD is still playing fine.
- On some devices there seems to be an audio hardware or 'in Android' buffer filling up. Despite the PTS values still being sync, Audio gets delayed, and when pausing, it still outputs audio for a second or two. 
  - It might be better to let the device pull audio instead of pushing audio to the device. I never had such a buffer issue, neither when I let the device pull (on Win/Mac/Linux),  nor on Android using QtMM.
  - Or there is a setting to keep the audio hardware or 'in Android' buffer very small, to ensure low latency.

**Comment by geminixdev** [comment]
> * On some devices there seems to be an audio hardware or 'in Android' buffer filling up. Despite the PTS values still being sync, Audio gets delayed, and when pausing, it still outputs audio for a second or two.
>   
>   * It might be better to let the device pull audio instead of pushing audio to the device. I never had such a buffer issue, neither when I let the device pull (on Win/Mac/Linux),  nor on Android using QtMM.
>   * Or there is a setting to keep the audio hardware or 'in Android' buffer very small, to ensure low latency.

I see it's using QAudioSink, so there is already 'pulling audio' in `readData`, and there is `setBufferSize()`. I will play with it to see if the issue can get addressed there.

**Comment by geminixdev** [comment]
> > * On some devices there seems to be an audio hardware or 'in Android' buffer filling up. Despite the PTS values still being sync, Audio gets delayed, and when pausing, it still outputs audio for a second or two.
> >   
> >   * It might be better to let the device pull audio instead of pushing audio to the device. I never had such a buffer issue, neither when I let the device pull (on Win/Mac/Linux),  nor on Android using QtMM.
> >   * Or there is a setting to keep the audio hardware or 'in Android' buffer very small, to ensure low latency.
> 
> I see it's using QAudioSink, so there is already 'pulling audio' in `readData`, and there is `setBufferSize()`. I will play with it to see if the issue can get addressed there.

Yes, limiting the buffer with setBufferSize() at the QAudioSink fixes that 'growing audiobuffer' problem on the devices which showed it. 

May I suggest to add a `setBufferSize(int value)` to `QAVAudioOutput`, analog to `setVolume(qreal v)` ?
This could set an `int bufferSize` varaible in `QAVAudioOutputPrivate`, which would then be used to set the buffer right before audioOutput->start(this), like this:

```
if (bufferSize != 0)
     audioOutput->setBufferSize(bufferSize); 
audioOutput->start(this);
```

16384 or 32768 seems to be a good value.

**Comment by valbok** [comment]
https://github.com/valbok/QtAVPlayer/pull/310

**Comment by valbok** [comment]
https://github.com/valbok/QtAVPlayer/pull/309 this should resolve issues with h264_mediacodec and missed frames

**Comment by geminixdev** [comment]
> #309 this should resolve issues with h264_mediacodec and missed frames

From what I've seen with h264_mediacodec it doesn't, there will still be frames missing:

The reason is that frames are missing **NOT** because there are multiple frames to be received after sending one packet with `avcodec_send_packet`. I have not noticed such a behavior with h264_mediacodec.

But there are frames missing because some packets don't get sent. Mediacodec isn't ready to accept sent packets, as long as there isn't first one frame (or more) in the receive buffer tried to be taken out with `avcodec_receive_frame`.

And whenever `avcodec_send_packet` fails (does not return 0) because mediacodec didn't accept a packet, and that happens often, then that packet gets lost, and with it the frame which it would (later) produce.

From my observations, and also as decribed in that blog post I mentioned above, after mediacodec rejects a packet (`avcodec_send_packet` does not return 0) it needs one or several frame receive attempts with `avcodec_receive_frame`, before that package can get successfully sent.

So you need to keep track of rejected packets and resubmit them.

Yes I know, it's ugly and a very different order than ffmepg ususal.

**Comment by valbok** [comment]
Will check any missing frames again, but did you try the latest changes?

**Comment by geminixdev** [comment]
> * On another armeabi-v7a device the GUI thread seems to get overloaded when playing Full HD, which is visible in the video not playing smoothly, despite decoding being fast. HD is still playing fine.

I noticed also, that on affected devices the video stutters when I do something in the GUI, hinting at the CPU being involved.

My _first guess_ is that here we have an image conversion / pixelformat conversion in CPU, happening. After the fast decoding in hardware the frame goes to QVideoFrame, but this time it's not as in software decoding in yuv420p (?), but in AV_PIX_FMT_MEDIACODEC format. So somewhere before pushing to display in the videosink (OpenGL?) it seems to get converted, using CPU, not accelerated.

No problem on the faster arm64-v8a devices, but apparently too much, at least at  higher resolutions for armeabi-v7a devices. 

How to verify and speed up?

**Comment by valbok** [comment]
Decoding is hw accelerated but rendering is not. Converting to QVideoFrame downloads data to CPU and converts to Format_YUV420P after that QML will use software rendering based on CPU.

**Comment by geminixdev** [comment]
> Will check any missing frames again, but did you try the latest changes?

I didn't, my comments where solely based on my previous tests where I chased the lost frames, trying to understand why avcodec_receive_frame wouldn't give them to me. Spending hours on it, until I tracked the 'packets in' too, and saw them getting rejected and lost - and saw that the frames missing should have been exactly the PTS values of the rejected (and not resubmitted) packets. 
This affected within 10 packets between 2 and 4 packets, prossibly depending on the b-frames used. 

Once I made sure to resubmit (prepend to the packet queue) every rejected packet there were no more lost frames.

**Comment by geminixdev** [comment]
> Decoding is hw accelerated but rendering is not. Converting to QVideoFrame downloads data to CPU and converts to Format_YUV420P after that QML will use software rendering based on CPU.

Thanks, that's what I thought. 

So either these devices need 
- hardware accelerated conversion to YUV420P or 
- not 'software rendering" but OpenGL,
- or both. Pushing the AV_PIX_FMT_MEDIACODEC frames directly to a surface usable in QVideoSink.

**Comment by valbok** [comment]
as I know Qt does not support mediacodec video frames, maybe opengl is possible but I've never seen yet how to convert AV_PIX_FMT_MEDIACODEC to OpenGL textures.

**Comment by geminixdev** [comment]
> as I know Qt does not support mediacodec video frames, maybe opengl is possible but I've never seen yet how to convert AV_PIX_FMT_MEDIACODEC to OpenGL textures.

I think they have it implemented somehow. On the same devices where I see the issues the QtMM based Player runs apparently fully accelerated, also in Full HD without any stuttering. 
Of course with all QtMM limitations and problems, but the acceleration part they must have somehow implemented.

**Comment by valbok** [comment]
> > Will check any missing frames again, but did you try the latest changes?
> 
> I didn't, my comments where solely based on my previous tests where I chased the lost frames, trying to understand why avcodec_receive_frame wouldn't give them to me. Spending hours on it, until I tracked the 'packets in' too, and saw them getting rejected and lost - and saw that the frames missing should have been exactly the PTS values of the rejected (and not resubmitted) packets. This affected within 10 packets between 2 and 4 packets, prossibly depending on the b-frames used.
> 
> Once I made sure to resubmit (prepend to the packet queue) every rejected packet there were no more lost frames.
Could you try this? It reports just expected and received frames, 
https://github.com/valbok/QtAVPlayer/pull/316

For me it shows 1440 reported and ~1431 received

**Comment by geminixdev** [comment]
> 
> Could you try this? It reports just expected and received frames, #316
> 
> For me it shows 1440 reported and ~1431 received

Will try to run tests later.

A first thought:

I cannot remember having seen code in QtAVPlayer to drain the codec at the end. From my tests, there are still about 1 - 5 frames in the mediacodec buffer, which you get out by sending a NULL packet and then avcodec_receive_frame until it returns AVERROR_EOF.

So some (1-5) of the missing frames you see might still be stuck in mediacodec (unless I did miss a draining happening in QtAVPlayer code).

Here the drain function I used in testing:

```
void QAVFrameCodec::drainAndFlush() const
{
    Q_D(const QAVCodec);
    if (!d->avctx)
        return;

    // Instead of valid input, send NULL in avcodec_send_packet()
    int ret = avcodec_send_packet(d->avctx, NULL);
    qWarning() << __FUNCTION__ << "avcodec_send_packet" << ret;

    AVFrame *frame = av_frame_alloc();

    while (true) {
        ret = avcodec_receive_frame(d->avctx, frame);
        qWarning() << __FUNCTION__ << "avcodec_receive_frame" << ret;

        if (ret == AVERROR_EOF)
            break;
    }

    av_frame_free(&frame);
}
```

Here I'm throwing the frames away, I did just run this for testing how many there will be left, and testing if it is needed to drain mediacodec before starting the next file/stream. (I's not, draining or not draining did not make a difference for the next file / stream decoding in my tests.)

**Comment by valbok** [comment]
I meant to build qml_example as is and run it and make sure that no missed frames.

**Comment by geminixdev** [comment]
> > Will check any missing frames again, but did you try the latest changes?
> 
> I didn't, my comments where solely based on my previous tests where I chased the lost frames, trying to understand why avcodec_receive_frame wouldn't give them to me. Spending hours on it, until I tracked the 'packets in' too, and saw them getting rejected and lost - and saw that the frames missing should have been exactly the PTS values of the rejected (and not resubmitted) packets. This affected within 10 packets between 2 and 4 packets, prossibly depending on the b-frames used.
> 
> Once I made sure to resubmit (prepend to the packet queue) every rejected packet there were no more lost frames.

You have successfully addressed and fixed this in #317 - tested and running fine in my test setup, playing streams for hours without a single frame missing. 

No more lost packets and as a consequence no more lost frames. 

Your solution is also more elegant than the code I presented above (the decode_mediacodec function posted several posts above). Excellent! Thanks!

**Comment by valbok** [comment]
Great, so idea is in:

    AVERROR(EAGAIN): input is not accepted in the current state - user must read output with avcodec_receive_frame()
    (once all output is read, the packet should be resent, and the call will not fail with EAGAIN)

So closing the issue? Thanks for your help, technically you made all job there.

**Comment by geminixdev** [comment]
> * On another armeabi-v7a device the GUI thread seems to get overloaded when playing Full HD, which is visible in the video not playing smoothly, despite decoding being fast. HD is still playing fine.

That armeabi-v7a device works fine with the latest code in #317 up to HD. For Full HD however it starts to reply for both `avcodec_send_packet` and `avcodec_receive_frame` with `AVERROR(EAGAIN)`, which is illegal according to ffmpeg specs. And it does that several times in a row, thus delaying the decoding too much to play smoothly.

From the ffmpeg docs (https://ffmpeg.org/doxygen/trunk/group__lavc__encdec.html):

> A codec is not allowed to return [AVERROR(EAGAIN) for both sending and receiving. This would be an invalid state, which could put the codec user into an endless loop. The API has no concept of time either: it cannot happen that trying to do [vcodec_send_packet()results in AVERROR(EAGAIN), but a repeated call 1 second later accepts the packet.

But this happens continuously at that device when decoding Full HD h264 streams, and 3-5 times in a row. It does not happen when decoding SD or HD, only Full HD.

Dunno what to do with that. That device plays Full HD quite fine with QtMM. It seems to me to be a bug on the ffmpeg or device side, but not in QtAVPlayer code.

Therefore yes, close it. Getting the part between decoding and displaying on screen hardware accelerated is another issue.

Thanks!

**Comment by valbok** [comment]
Have you tried QtMultimedia with ffmpeg and seen no issues with FullHD?

**Comment by geminixdev** [comment]
> Have you tried QtMultimedia with ffmpeg and seen no issues with FullHD? By default QtMM uses builtin MediaPlayer

My original solution is based on on QtMM with native Android builtin MediaPlayer, in Qt5. There the tested devices all work hardware accelerated, playing smoothly also in full HD.

In Qt6 'QtMM with ffmpeg' was not working because QtMM ffmpeg in Qt6 supports playing from local files, but not from network resources, no http:// .. or https://... urls. Waiting for Qt to release Qt 6.5.1 with the fix, see https://bugreports.qt.io/browse/QTBUG-111378

Nevertheless, a real ffmpeg based solution is much preferred over a QtMM black box solution with ffmpeg somewhere used internally. When playing damaged streams it's easier to control and recover if there is direct access to PTS and frames and no black box. (I have experience with playing video on Windows / mac / Linux using the ffmpeg libraries directly, not QtMM, and there I can control the handling of bad streams.) Also I prefer to play from buffer (QIODevice) instead of an Url, which is possible with ffmpeg, but not really with QtMM. Hence my excitement for QtAVPlayer.
